### PR TITLE
Demo Avro for FlightMap serialization. [not for review]

### PR DESCRIPTION
**Pull request by jmczerk**
Avro equivalent of @wchamber protobuf demo.  

Avro schema evolution rules: https://avro.apache.org/docs/1.10.2/spec.html#Schema+Resolution

### CODE REVIEW

File: src/test/java/bio/terra/common/iam/AuthenticatedUserRequestTest.java [hunk 1]
```diff
@@ -119,4 +125,108 @@ public void testVectors() throws JsonProcessingException {
...
+    AuthenticatedUserRequest request =
+        AuthenticatedUserRequest.builder()
+            .setEmail(EMAIL_ADDRESS)
+            .setSubjectId(SUBJECT_ID)
+            .setToken(TOKEN)
+            .build();
+
+    FlightMap inMap = request.putIn(new FlightMap(), "foo");
+    FlightMap outMap = new FlightMap();
+    outMap.fromJson(inMap.toJson());
```
Comment (wchamber) [inline src/test/java/bio/terra/common/iam/AuthenticatedUserRequestTest.java:140]: we should both update to also use the test fixture for stairway serialization added in https://github.com/DataBiosphere/stairway/pull/64 to check for input FlightMap compatibility.

File: src/main/java/bio/terra/common/iam/AuthenticatedUserRequest.java [hunk 2]
```diff
@@ -40,6 +58,95 @@ public String getToken() {
...
+   * illustrate.
+   */
+  @VisibleForTesting
+  static <T> FlightMap putRecord(
+      FlightMap flightMap, String key, Schema schema, T record, Class<T> type) {
+    try {
+
+      // Encode the object with the "writer" schema into the byte stream.
+      OutputStream stream = new ByteArrayOutputStream();
+      Encoder jsonEncoder = EncoderFactory.get().jsonEncoder(schema, stream);
```
Comment (wchamber) [inline src/main/java/bio/terra/common/iam/AuthenticatedUserRequest.java:91]: any particular reason to use the json encoder instead of the binary encoder?

Comment (jmczerk) [reply src/main/java/bio/terra/common/iam/AuthenticatedUserRequest.java:91]: As one would expect, binary encoded data looks like a lot like protobuf:
```
 test@example.com\u000ESubject 0123.456-789AbCd\u0000\u0006foo
```

The big downside here is that "writer schema" must be stored with the data, and is JSON only; Avro is designed to be used either with large data sets where the schema is stored in a header with many records, or using a [schema registry](https://www.confluent.io/product/confluent-platform/data-compatibility) where a fingerprint is used as a key.  We could do our own schema management, but this seems like a lot of added complexity, and kind of moves backwards on the tight-coupling we're trying to get away from (assuming the schemas get stored in the database).

On the positives, I do like the no-code-generation option (see other PR); if we were to go with Avro with code generation I suspect there's a bit of work we'd want to do to tweak the [Velocity templates](https://velocity.apache.org/) that Avro uses to generate the classes

Comment (jmczerk) [reply src/main/java/bio/terra/common/iam/AuthenticatedUserRequest.java:91]: Btw, there may also be no-code-generation options for protobuf: https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/DynamicMessage

Comment (wchamber) [reply src/main/java/bio/terra/common/iam/AuthenticatedUserRequest.java:91]: Ha, I actually really dislike the no-code-generation options. Give me generated classes and type safety where I can't fat finger the field name.

Comment (jmczerk) [reply src/main/java/bio/terra/common/iam/AuthenticatedUserRequest.java:91]: In general I agree, but a bad object mode for the generated code can be a downside, which I think is reflected in the semantics of interaction with FlightMap here.

Protobuf seems to have been pretty thoughtfully designed their Java object model.  Avro can achieve something similar, but would require more customization to get there.  Out of the box the FlightMap-facing interface is awkward.

**Pull request closed without merging**